### PR TITLE
Amar weather-app w3

### PR DESCRIPTION
**Pull request by Amar-Mahdy**

### CODE REVIEW

File: project/hackyourweather/src/components/WeatherList.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      {errMsg && <h3 className="errorMessage">{errMsg}</h3>}
+      {isLoading && <h3 className="errorMessage">Loading...</h3>}
+
+      {data.length > 0 ? (
+        <div className="container">
+          <h4 className="click">Click on the city to see the charts</h4>
+          {data &&
+            data.map((city, index) => {
+              return (
+                <Card key={index} city={city} handleDelete={handleDelete} />
```
Comment (crevulus) [inline project/hackyourweather/src/components/WeatherList.js:53]: good, you caught the "key" error from last time. Just FYI, it is best **not** to use `index` for a key, but instead to use a unique identifier from the data. 🟢

File: project/hackyourweather/package.json [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    "bootstrap": "^5.1.3",
+    "msw": "^0.35.0",
+    "react": "^17.0.2",
+    "react-bootstrap": "^2.0.2",
+    "react-dom": "^17.0.2",
+    "react-icons": "^4.3.1",
+    "react-router-dom": "^6.0.2",
+    "react-scripts": "4.0.3",
+    "recharts": "^2.1.6",
+    "uuid": "^8.3.2",
```
Comment (crevulus) [inline project/hackyourweather/package.json:18]: I see lots of packages here that you're not using. `uuid`, `whatwg-fetch`, `bootstrap`. Be sure to remove them from your `package.json` and `uninstall` via `npm`. 🟢

File: project/hackyourweather/src/components/forecast.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import React from "react";
+import { getCelcius } from "../converter/converter";
+
+const forecast = ({ city }) => {
+  const { coord } = city;
+  const cityLocation = `${coord.lat},${coord.lon}`;
+  return (
+    <>
+      <span>
+        <strong>min temp: </strong> {getCelcius(city.main.temp_min)}
```
Comment (crevulus) [inline project/hackyourweather/src/components/forecast.js:10]: You might want to check this method: your app currently says London is max temp **-260.0**! I know it's cold there but not that cold 🔴

File: project/hackyourweather/src/components/card.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import MainForecast from "./MainForecast.js";
+import { ImCancelCircle } from "react-icons/im";
+
+const card = ({ city, handleDelete }) => {
+  return (
+    <div className="card">
+      <button className="clear-btn" onClick={() => handleDelete(city.id)}>
+        <ImCancelCircle />
+      </button>
+      <Header city={city} />
```
Comment (crevulus) [inline project/hackyourweather/src/components/card.js:13]: Good component-isation here

File: project/hackyourweather/src/components/MainForecast.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import React from "react";
+
+const MainForecast = ({ city }) => {
+  return (
+    <>
+      <h4>
```
Comment (crevulus) [inline project/hackyourweather/src/components/MainForecast.js:6]: I mentioned this in my last review. Please correct it! https://github.com/HackYourHomework/React-class33/pull/8#discussion_r744119866 🔴

File: project/hackyourweather/src/components/MainForecast.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+        {city.main}
+        <br />
+        <strong>{city.description}</strong>
```
Comment (crevulus) [inline project/hackyourweather/src/components/MainForecast.js:9]: I mentioned this, too, but left it open: https://github.com/HackYourHomework/React-class33/pull/8#discussion_r744119918
The problem here is that the description should have a capital letter. The data you receive is all lower-case letters, so you need to find a way to convert the first letter to upper case. It's a small difference, but it's very important that your websites match client expectations! 🔴

File: project/hackyourweather/src/components/forecast.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+      </span>{" "}
+      <br />
+      <span>
+        <strong>max temp: </strong>
+        {getCelcius(city.main.temp_max)}
+      </span>{" "}
```
Comment (crevulus) [inline project/hackyourweather/src/components/forecast.js:16]: What is this empty space for? 🟢

File: project/hackyourweather/src/components/form.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+function Form({ handleCity, handleSubmit, cityName }) {
+  return (
+    <>
+      <form onSubmit={handleSubmit}>
+        <label htmlFor="search">search for city's weather</label>
+        <input
+          type="text"
+          className="form-control"
+          placeholder="e.g Groningen"
```
Comment (crevulus) [inline project/hackyourweather/src/components/form.js:11]: Good UX

File: project/hackyourweather/src/components/form.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+          onChange={handleCity}
```
Comment (crevulus) [inline project/hackyourweather/src/components/form.js:12]: your `handleCity` function in the parent just sets a global state variable (context api) -- do you really need to write it in the parent and pass it down here? Wouldn't it be better to set it here directly, where it's used? 🔴

File: project/hackyourweather/src/components/cityPage.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
+import React, { useEffect, useState } from "react";
+import { useNavigate } from "react-router-dom";
```
Comment (crevulus) [inline project/hackyourweather/src/components/cityPage.js:2]: if you have two imports from the same module, put them in the same import statement.🟠

Comment (crevulus) [reply project/hackyourweather/src/components/cityPage.js:2]: I haven't even used react router v6 yet - very cool!

File: project/hackyourweather/src/components/cityPage.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+      try {
+        const response = await fetch(
+          `https://api.openweathermap.org/data/2.5/forecast?id=${cityId}&units=metric&appid=${process.env.REACT_APP_OPENWEATHERMAP_API_KEY}`
+        );
+        const data = await response.json();
+        console.log(data);
+        setData(data);
+        fillData(data);
+      } catch (err) {
+        console.log(err);
```
Comment (crevulus) [inline project/hackyourweather/src/components/cityPage.js:33]: Could you communicate these errors to the user somehow? 🟠

File: project/hackyourweather/src/components/cityPage.js [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      });
+    }
+    setChart(tempData);
+  };
+
+  return (
+    <div>
+      {chart.length !== 0 && (
+        <>
+          <button className="back" onClick={() => navigate("/")}>
```
Comment (crevulus) [inline project/hackyourweather/src/components/cityPage.js:54]: `useNavigate` has a specific method for going back - that's basically why it was added to react-router v6! Better to use that: https://dev.to/salehmubashar/usenavigate-tutorial-react-js-aop 🟢

File: project/hackyourweather/src/components/cityPage.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+            Back
+          </button>
+          <h1 className="container">charts of 5 days for your favorite city</h1>
+          <h3>
+            {data.city.name}, {data.city.country}
+          </h3>
+          <ResponsiveContainer width={1300} height={500}>
```
Comment (crevulus) [inline project/hackyourweather/src/components/cityPage.js:61]: This is some advanced stuff!

**Comment by crevulus** [comment]
Hi Amar,

This is a vast improvement from week 1! It seems like you're really getting to grips with React very quickly. I've left a lot of compliments in my review and I'm very impressed.

As for things that are missing, the main one is tests! I know the requirements to have "every component tested" is no longer a requirement, but I don't see any tests at all. Please include some basic tests for a few of your components and/or some of your logic, functions etc.

And there's also some things you've left in there since week one that should have already been changed. Look out for them in my review.

**Review by crevulus — CHANGES_REQUESTED** [review]

File: project/hackyourweather/src/converter/converter.js [hunk 14]
```diff
@@ -0,0 +1,3 @@
+export const getCelcius = (kelvin) => {
```
[the remark this answers is not in the document]
Comment (Amar-Mahdy) [reply project/hackyourweather/src/converter/converter.js:1]: thanks you so much for your nice impression, i appreciate it very well.
and yes utils is a descriptive name

File: project/hackyourweather/src/App.css [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+}
+.chart {
+  text-align: center;
+  margin-bottom: 30px;
+}
+
+.card {
+  padding: 0px 0px 20px 20px;
+  border: 2px solid black;
+  width: 30%;
```
[the remark this answers is not in the document]
Comment (Amar-Mahdy) [reply project/hackyourweather/src/App.css:16]: not actually, but i probably need to do it more responsive

File: project/hackyourweather/src/components/WeatherContext.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
+import react, { useState, createContext } from "react";
+
+export const WeatherContext = createContext();
+
+export const WeatherProvider = (props) => {
+  const [data, setData] = useState([]);
+  const [cityName, setCityName] = useState("");
+  const [isLoading, setIsLoading] = useState(false);
```
[the remark this answers is not in the document]
Comment (Amar-Mahdy) [reply project/hackyourweather/src/components/WeatherContext.js:8]: maybe to false ?

File: project/hackyourweather/src/components/form.js (hunk 9 above)
Comment (Amar-Mahdy) [reply project/hackyourweather/src/components/form.js:12]: that,s crazy, your right, why do i need it out there

File: project/hackyourweather/src/components/cityPage.js (hunk 10 above)
Comment (Amar-Mahdy) [reply project/hackyourweather/src/components/cityPage.js:2]: what do you mean by the first comment?

File: project/hackyourweather/src/components/WeatherContext.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (Amar-Mahdy) [reply project/hackyourweather/src/components/WeatherContext.js:8]: yes i forgot it

File: project/hackyourweather/src/components/WeatherContext.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+      console.log(error);
+    } finally {
+      setIsLoading(false);
+      setCityName("");
+    }
+  };
+
+  return (
+    <WeatherContext.Provider
+      value={[
```
Comment (crevulus) [inline project/hackyourweather/src/components/WeatherContext.js:34]: This is not the correct way to structure the `value` prop. It should be an object, i.e. `value={{ // values in here }}`. Wihtout this, you cannot use object destructuring when you try to access context values. https://www.freecodecamp.org/news/array-and-object-destructuring-in-javascript/ 🔴

Comment (crevulus) [reply project/hackyourweather/src/components/WeatherContext.js:34]: you will have to change the way you destructure on line 8 in `WeatherList`, too.

**Review by crevulus — CHANGES_REQUESTED** [review]
What a beautiful website! There's some very advanced stuff in there, and you've paid close attention to creating beautiful visuals. Well done!

Your code is a bit messy in places. You're passing state across components rather than using your Context provider as your single source of truth, you've got some functions that do essentially the same thing, and your renaming things is unnecessary. That's causing errors in your UI that are hard to track.

There's also one key thing missing: tests! The requirement for tests is definitely smaller, but that doesn't mean 0 tests. Please include some tests with your next commit.

**Comment by Amar-Mahdy** [comment]
updating the reviews

**Pull request closed without merging**